qutebrowser/misc/notree.py

```diff
@@ -0,0 +1,287 @@
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Tree library for tree-tabs.
+
+The fundamental unit is the Node class.
+
+Create a tree with with Node(value, parent):
+root = Node('foo')
+child = Node('bar', root)
+child2 = Node('baz', root)
+child3 = Node('lorem', child)
+
+You can also assign parent after instantiation, or even reassign it:
+child4 = Node('ipsum')
+child4.parent = root
+
+Assign children:
+child.children = []
+child2.children = [child4, child3]
+child3.parent
+> Node('foo/bar/baz')
+
+Render a tree with render_tree(root_node):
+render_tree(root)
+
+> ('', 'foo')
+> ('├─', 'bar')
+> ('│ ├─', 'lorem')
+> ('│ └─', 'ipsum')
+> ('└─', 'baz')
+"""
+import enum
+import typing
+
+# For Node.render
+CORNER = '└─'
+INTERSECTION = '├─'
+PIPE = '│'
+
+
+class TreeError(RuntimeError):
+    """Exception used for tree-related errors."""
+
+
+class TraverseOrder(enum.Enum):
+    """To be used as argument to traverse().
+
+    Implemented orders are pre-order and post-order.
+    Attributes:
+        PRE: pre-order (parents before children). Same as in Node.render
+        POST: children of a node are always yield before their parent.
+        POST_R: Like POST, but children are yield in reverse order
+    """
+
+    PRE = 'pre-order'
+    POST = 'post-order'
+    POST_R = 'post-order-reverse'
+
+
+class Node():
+    """Fundamental unit of notree library.
+
+    Attributes:
+        value: The element (ususally a tab) the node represents
+        parent: Node's parent.
+        children: Node's children elements.
+        siblings: Children of parent node that are not self.
+        path: List of nodes from root of tree to self
+    value, parent, and children can all be set by user. Everything else will be
+    updated accordingly, so that if `node.parent = root_node`, then `node in
+    root_node.children` will be True.
+    """
+
+    sep: str = '/'
+    __parent: typing.Optional['Node'] = None
+    __uid: int
+    __last_uid: int = 0  # this is a global 'static' class attribute
+
+    PARENT_TYPE = typing.TypeVar('PARENT_TYPE')
+
+    def __init__(self,
+                 value: PARENT_TYPE,
+                 parent: typing.Optional['Node'] = None,
+                 childs: typing.Sequence['Node'] = (),
+                 uid: typing.Optional[int] = None) -> None:
+        if uid is not None:
+            self.__uid = uid  # TODO find a way to check if given ID exists already
+        else:
+            self.__uid = self.__get_new_uid()
+
+        self.value = value
+        # set initial values so there's no need for AttributeError checks
+        self.__parent = None  # type: typing.Optional['Node']
+        self.__children = []  # type: typing.List['Node']
+
+        # For render memoization
+        self.__modified = False
+        self.__set_modified()  # not the same as line above
+        self.__rendered = []  # type: typing.List[typing.Tuple[str, 'Node']]
```

Comment:
This should be a tuple instead of a list. Since we're returning it, we don't want anyone changing it (while they can change nodes still, this is one area of confusion that can be fixed). Also - tuples use memory than lists do.

You can still use a list to build it up to keep the logic the same, but return a tuple in the end.